nexus/db-queries/src/db/datastore/external_ip.rs

```diff
@@ -281,4 +376,167 @@ impl DataStore {
             .await
             .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
     }
+
+    /// Fetch all Floating IP addresses for the provided project.
+    pub async fn floating_ips_list(
+        &self,
+        opctx: &OpContext,
+        authz_project: &authz::Project,
+        pagparams: &PaginatedBy<'_>,
+    ) -> ListResultVec<FloatingIp> {
+        use db::schema::floating_ip::dsl;
+
+        opctx.authorize(authz::Action::ListChildren, authz_project).await?;
+
+        match pagparams {
+            PaginatedBy::Id(pagparams) => {
+                paginated(dsl::floating_ip, dsl::id, &pagparams)
+            }
+            PaginatedBy::Name(pagparams) => paginated(
+                dsl::floating_ip,
+                dsl::name,
+                &pagparams.map_name(|n| Name::ref_cast(n)),
+            ),
+        }
+        .filter(dsl::project_id.eq(authz_project.id()))
+        .filter(dsl::time_deleted.is_null())
+        .select(FloatingIp::as_select())
+        .get_results_async(&*self.pool_connection_authorized(opctx).await?)
+        .await
+        .map_err(|e| public_error_from_diesel(e, ErrorHandler::Server))
+    }
+
+    /// Delete a Floating IP, verifying first that it is not in use.
+    pub async fn floating_ip_delete(
+        &self,
+        opctx: &OpContext,
+        authz_fip: &authz::FloatingIp,
+        db_fip: &FloatingIp,
+    ) -> DeleteResult {
+        use db::schema::external_ip::dsl;
+
+        // Verify this FIP is not attached to any instances/services.
+        if db_fip.parent_id.is_some() {
```

Comment:
Should getting the `db_fip` value and acting on the value for the detach action below be done in a transaction to ensure that we can't get a race condition between attaching and deleting?

Reply:
I think it would be improved by doing so, yes. The actual `diesel::update` attach/detach/delete operations all include a filter on parent_id, so currently if preempted by another write they will back out without doing anything unintended, but I'd prefer a transaction too.

Reply:
The [`update_and_check()`](https://github.com/oxidecomputer/omicron/blob/bcd7ac5d5e10355774310aa583aa59253eb880e5/nexus/db-queries/src/db/update_and_check.rs#L1) machinery is probably going to be helpful here. It lets one do conditional updates, say on the parent ID being NULL, and returning a detectable error in the case where that condition is not met.